lib/build.py

```diff
@@ -1097,11 +1105,12 @@ def prepare(self, miss_ct):
       if (str(self.image.ref) == str(self.base_image.ref)):
          ch.FATAL("output image ref same as FROM: %s" % self.base_image.ref)
       # Close previous stage if needed.
-      if (miss_ct == 0 and self.image_i > 0):
-         # While there haven’t been any misses so far, we do need to check out
-         # the previous stage (a) to read its metadata and (b) in case there’s
-         # a COPY later. This will still be fast most of the time since the
-         # correct branch is likely to be checked out already.
+      if (        self.image_i > 0
+          and not isinstance(bu.cache, bu.Disabled_Cache)):
```

Comment:
Need to figure out why `Disabled_Cache` fails rather than just excluding it.

Reply:
I should note that this is different from, and doesn't interfere with copying `stage_0` to `stage_1`, which occurs later.

Reply:
Sounds like the logic within `Disabled_Cache` needs to be fixed somehow then?

Reply:
> Sounds like the logic within Disabled_Cache needs to be fixed somehow then?

Hmm, perhaps. FWIW, I did consider it.

It appears to me that the issue is that the `FROM` instruction subclass needs to behave differently when the build cache is disabled. The build cache class is designed to be instruction agnostic, the methods behave a certain way based on the cache mode, not the calling Dockerfile instruction.

Recall that the FROM instruction calls it's parent instruction's checkout method; the parent could be any one of the other 16 instruction subclasses, some of which make no changes to the image directory at all. This is why FROM needs to know if the build cache is enabled or not, prior to calling a parent's checkout -- it only needs to when the cache is enabled.

If you have ideas, let me know. I spent quite a bit of time here but I certainly may have missed something.

Reply:
OK. Let me send it back to myself for review and I'll see if I can think of something.